pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,644 @@
+        )
+
+        self.resource_group = self._create_resource_group()
+        self.base_tag = tag
+
+    def _create_network_security_group(self):
+        """Create a network security group.
+
+        This method creates a network security groups that allows the user
+        to ssh into the machine and execute commands.
+
+        Returns:
+            The network security object created by Azure
+
+        """
+        security_group_name = "{}-sgn".format(self.tag)
+        nsg_group = self.network_client.network_security_groups
+
+        self._log.debug('Creating Azure network security group')
+        nsg_call = nsg_group.create_or_update(
+            resource_group_name=self.resource_group.name,
+            network_security_group_name=security_group_name,
+            parameters={
+                "location": self.location,
+                "security_rules": [
+                    {
+                        "name": "SSH",
+                        "properties": {
+                            "priority": 300,
+                            "protocol": "TCP",
+                            "access": "Allow",
+                            "direction": "Inbound",
+                            "sourceAddressPrefix": "*",
+                            "sourcePortRange": "*",
+                            "destinationAddressPrefix": "*",
+                            "destinationPortRange": "22"
+                        }
+                    }
+                ]
+            }
+        )
+
+        return nsg_call.result()
+
+    def _create_resource_group(self):
+        """Create a resource group.
+
+        This method creates an Azure resource group. Every other componet that
+        we create will be contained into this resource group. This means that
+        if we delete this resource group, we delete all resources associated
+        with it.
+
+        Returns:
+            The resource group created by Azure
+
+        """
+        resource_name = "{}-rg".format(self.tag)
+        self._log.debug('Creating Azure resource group')
+
+        return self.resource_client.resource_groups.create_or_update(
+            resource_name,
+            {
+                "location": self.location
+            }
+        )
+
+    def _create_virtual_network(self, address_prefixes=None):
+        """Create a virtual network.
+
+        This method creates an Azure virtual network to be used
+        when provisioning a subnet.
+
+        Args:
+            address_prefixes:  list of strings, A list of address prefixes
+                               to be used in this virtual network.
+        Returns:
+            The virtual network created by Azure
+
+        """
+        if address_prefixes is None:
+            address_prefixes = ["10.0.0.0/16"]
+
+        virtual_network_name = "{}-vnet".format(self.tag)
+
+        self._log.debug('Creating Azure virtual network')
+        network_call = self.network_client.virtual_networks.create_or_update(
```

Comment:
Same question for any resource we create or update, we should probably add tags if we can.

Reply:
Added tags to all resources